src/util.ts

```diff
@@ -209,3 +210,15 @@ export function varName(s: string): string {
   // Add _ if the string has leading numbers.
   return (s.match(/^\d+/) ? '_' : '') + alphanumericS;
 }
+
+export function logicalExpr<T>(op: LogicalOperand<T>, cb: Function): string {
+  if (isLogicalNot(op)) {
+    return '!' + logicalExpr(op.not, cb);
```

Comment:
I'm not sure additional parentheses are needed. [I added](https://github.com/vega/vega-lite/blob/7c3d1e42e4976be023035182a407bb2fa042690a/test/compile/selection/predicate.test.ts#L49) your `!(A && B)` example to the tests and it produces what I would expect. I think we only need parentheses around the array operators to group all those items together.